src/server/game/Movement/PathGenerator.cpp

```diff
@@ -915,35 +915,121 @@ void PathGenerator::ReducePathLenghtByDist(float dist)
     if (_pathPoints.size() < 2) // path building failure
         return;
 
-    uint32 i = _pathPoints.size();
-    G3D::Vector3 nextVec = _pathPoints[--i];
-    while (i > 0)
+    /*
+    evolution of idx:
+     x---->x---->x---->X
+     0     1     2     3
+                idx   idx+1
+          idx  idx+1
+    idx  idx+1
+    ---------------------------
+    diffVec when idx==1 in example above:
+            ---->
+    */
+    int idx = _pathPoints.size() - 2;
+    float totalLengthSoFar = 0.0f;
+    G3D::Vector3 diffVec;
+    for ( ; idx >= 0; idx--)
     {
-        G3D::Vector3 currVec = _pathPoints[--i];
-        G3D::Vector3 diffVec = (nextVec - currVec);
-        float len = diffVec.length();
-        if (len > dist)
-        {
-            float step = dist / len;
-            // same as nextVec
-            _pathPoints[i + 1] -= diffVec * step;
-            _sourceUnit->UpdateAllowedPositionZ(_pathPoints[i + 1].x, _pathPoints[i + 1].y, _pathPoints[i + 1].z);
-            _pathPoints.resize(i + 2);
+        diffVec = _pathPoints[idx + 1] - _pathPoints[idx];
+        totalLengthSoFar += diffVec.length();
+        if (totalLengthSoFar >= dist)
             break;
+    }
+
+    if (totalLengthSoFar > dist)
+    {
+        float delta = totalLengthSoFar - dist;
+        G3D::Vector3 newFinalPoint = _pathPoints[idx] + diffVec.unit() * delta;
+        _sourceUnit->UpdateAllowedPositionZ(newFinalPoint.x, newFinalPoint.y, newFinalPoint.z);
+
+        /*
+        3 potential problems could have occured up until this point:
+        1) the "actual end position" (closest position to the target found by the PathGen) could be not exactly on the target.
+        Which means the distance from the target to the last point of the path BEFORE THE CUT could be > 0
+        and add into that the cutted distance and this can result in the target being further away than dist to the 
+        path's end position AFTER CUT.
+        2) AFTER THE CUT, the Z of the new last point is normalized 
+        (to make sure our point is on the ground in case non flying/swimming unit).
+        This means the distance from the target to the path's last point AFTER CUT can be higher than dist.
+        3) The path's last point AFTER CUT can be out of LoS from the target.
+
+        We need to make sure that in ANY case: 
+        a) the last point after cut is at MAX at "dist" yards from the target + marginal error factor (0.5f)
```

Comment:
actually you have not guarantee that a valid path could ever be at MAX at dist, i.e. PATHFIND_INCOMPLETE case which by definition doesn't reach the destination